plugin/evm/vm.go

```diff
@@ -1439,6 +1454,50 @@ func (vm *VM) getAtomicTx(txID ids.ID) (*Tx, Status, uint64, error) {
 	}
 }
 
+// getAtomicTxFromProcessingBlocks takes in a [txID] and searches for a matching atomic tx in processing blocks
+// Specifically we look from blocks with lastAcceptedBlockHeight + 1  to preferredBlockHeight.
+// This function assumes that we have already searched accepted blocks for atomic txs.
+// We return an error if we cannot find a matching atomic tx in the processing blocks.
+func (vm *VM) getAtomicTxFromProcessingBlocks(txID ids.ID) (*Tx, uint64, error) {
+	return getAtomicTxFromProcessingBlocks(
+		txID,
+		vm.blockChain.CurrentBlock().Number.Uint64(),
+		vm.blockChain.LastAcceptedBlock().Header().Number.Uint64(),
+		vm.chainConfig.IsApricotPhase5,
+		vm.blockChain.GetBlockByNumber,
+		vm.codec,
+	)
+}
+
+func getAtomicTxFromProcessingBlocks(
```

Comment:
an alternate implementation idea would be tracking the pending txs in the atomic backend (eg in a map), then removing them when the block is accepted/rejected.
https://github.com/ava-labs/coreth/blob/master/plugin/evm/atomic_backend.go#L364
https://github.com/ava-labs/coreth/blob/master/plugin/evm/atomic_state.go#L49
https://github.com/ava-labs/coreth/blob/master/plugin/evm/atomic_state.go#L91

not sure if it is simpler

Reply:
Thanks for the suggestion! I chose to iterate through the blocks [last accepted +1, preferred block] since it was obvious to me to look through those as the processing blocks and there were preexisting functions to check if those blocks have atomic txs. I'm looking more at the atomic backend since this looks good too, I see that `InsertTxs` is called in `onExtraStateChange` which is called in `Finalize`. Where is this being replicated in `FinalizeAndAssemble`?

Reply:
https://github.com/ava-labs/coreth/blob/37c9cc40b2bc8a4bc72906c4091a8839a888d4f7/plugin/evm/vm.go#L941

Reply:
It's not, since the block will be verified again after it's built.
We used to do this when we were planning to include the root of the atomic trie in the header https://github.com/ava-labs/coreth/commit/7b260fcf708cbf98c772dfa5575c67aef975f6d2

Reply:
Can we switch to @darioush 's suggestion of tracking the processing transactions in the atomic backend?

This will borrow from an existing abstraction where tracking the processing transactions is very easy as opposed to adding an additional getter to maintain for this one-off case.